### PR TITLE
[WOR-187] Add "fake" billing profile manager DAO

### PR DESCRIPTION
**Pull request by aherbst-broad**
## Context
[Relevant ticket](https://broadworkbench.atlassian.net/browse/WOR-187)

In order to create Azure workspaces, an azure billing profile (aka billing project in rawls) must be available. Rather than formally introduce the notion of Azure billing information to Rawls, the plan is to have the [terra-billing-profile-manager](https://github.com/dataBiosphere/terra-billing-profile-manager) be the source of truth for these records.

Until that service is ready and to unblock downstream integration + UI efforts for the workspaces team, we'll be introducing a "fake" billing profile manager in Rawls that will serve a hardcoded azure billing profile. Access to this billing profile will be gated by the same SAM permissions required for any other billing profile, as well as membership in the Azure alpha testing group.

## This PR
* Adds a `BillingProfileManagerDAO` class that returns a list of billing profiles for the user
* Integrates the `BillingProfileManagerDAO` with the `UserService`, specifically the `listBillingProfilesV2` call. In my testing this was the call that needs modification to get the fake profile into the dropdown on the createWorkspace dialog.
* Adds tests for the new DAO + unit level coverage for the existing `listBillingProfilesV2` method

----------

- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check that the **Product Owner** has signed off on any user-facing changes
- [ ] **Submitter**: Make sure Swagger is updated if API changes
  - [ ] **...and Orchestration's Swagger too!**
- [ ] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Database checks:
  * If PR includes new or changed db queries, include the explain plans in the description
  * Make sure liquibase is updated if appropriate
  * If doing a migration, take a backup of the
  [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
  and
  [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
  DBs in Google Cloud Console
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] Tell your tech lead (TL) that the PR exists if they want to look at it
- [ ] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via Slack and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

**Comment by aherbst-broad** [comment]
jenkins retest

File: core/src/main/scala/org/broadinstitute/dsde/rawls/billing/BillingProfileManagerDAO.scala [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    val azureConfig = config.azureConfig match {
+      case None =>
+        logger.warn("Multicloud workspaces enabled but no azure config setup, returning empty list of billng profiles")
+        return Future.successful(Seq())
+      case Some(value) => value
+    }
+
+    // NB until the BPM is live, we are returning a hardcoded
+    // Azure billing profile, with access enforced by SAM
+    samDAO.userHasAction(
```
Comment (aherbst-broad) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/billing/BillingProfileManagerDAO.scala:52]: There is an additional check that the UserService makes here https://github.com/broadinstitute/rawls/blob/ca50cb17a578da33abb4a1bb882fda940580ed56/core/src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala#L152 that should likely be factored into this class. It's fairly well entangled up with the rest of the code in that class that I decided to leave it there and defer until we are further down the road with the BPM integration.

File: core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/DataAccessSpec.scala [hunk 2]
```diff
@@ -67,7 +67,7 @@ class DataAccessSpec extends TestDriverComponentWithFlatSpecAndMatchers with Sca
         // can't use runAndWait here because we need to use our altDataSource
 
         val roundtripCheckFutures = roundtripCheckActions map { a => altDataSource.inTransaction { _ => a } }
-        Future.sequence(roundtripCheckFutures).futureValue(timeout(Span(10, Seconds)))
+        Future.sequence(roundtripCheckFutures).futureValue(timeout(Span(20, Seconds)))
```
Comment (aherbst-broad) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/DataAccessSpec.scala]: Unrelated, this test flaked for me in CI due to a timeout, bumping here to a more generous timeout.

File: core/src/test/scala/org/broadinstitute/dsde/rawls/user/UserServiceSpec.scala [hunk 3]
```diff
@@ -764,4 +773,78 @@ class UserServiceSpec extends AnyFlatSpecLike with TestDriverComponent with Mock
       actual.errorReport.statusCode.get shouldEqual StatusCodes.NotFound
     }
   }
+
+  behavior of "listBillingProjectsV2"
```
Comment (aherbst-broad) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/user/UserServiceSpec.scala:777]: We had no unit level coverage of the existing method, this is adding some.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/billing/BillingProfileManagerDAO.scala [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+/**
+ * Facade over the billing profile manager service. This service will eventually by the source of truth for
+ * billing profiles in the Terra system. For now, we are using this to layer in "external" billing profiles
+ * for the purposes of testing Azure workspaces.
+ */
+class BillingProfileManagerDAO(samDAO: SamDAO, config: MultiCloudWorkspaceConfig) extends LazyLogging {
+
+  /**
+   * Fetches the billing profiles to which the user has access
```
Comment (aherbst-broad) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/billing/BillingProfileManagerDAO.scala]: Update comment to express that this is limited to azure bps

File: core/src/main/scala/org/broadinstitute/dsde/rawls/billing/BillingProfileManagerDAO.scala [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+    val azureConfig = config.azureConfig match {
+      case None =>
+        logger.warn("Multicloud workspaces enabled but no azure config setup, returning empty list of billing profiles")
+        return Future.successful(Seq())
+      case Some(value) => value
+    }
+
+    for {
+      samUserResources <- samDAO.listUserResources(
```
Comment (aherbst-broad) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/billing/BillingProfileManagerDAO.scala]: Consider parametrizing this

File: core/src/main/scala/org/broadinstitute/dsde/rawls/billing/BillingProfileManagerDAO.scala [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+import org.broadinstitute.dsde.rawls.model.{CreationStatuses, RawlsBillingProject, RawlsBillingProjectName, RawlsBillingProjectResponse, SamResourceAction, SamResourceTypeNames, SamUserResource, UserInfo, AzureManagedAppCoordinates}
+
+import scala.concurrent.{ExecutionContext, Future}
+
+/**
+ * Facade over the billing profile manager service. This service will eventually by the source of truth for
+ * billing profiles in the Terra system. For now, we are using this to layer in "external" billing profiles
+ * for the purposes of testing Azure workspaces.
+ */
+class BillingProfileManagerDAO(samDAO: SamDAO, config: MultiCloudWorkspaceConfig) extends LazyLogging {
```
Comment (aherbst-broad) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/billing/BillingProfileManagerDAO.scala]: Will this work with terra ui billing project sharing?

File: core/src/test/scala/org/broadinstitute/dsde/rawls/user/UserServiceSpec.scala (hunk 3 above)
Comment (cahrens) [reply core/src/test/scala/org/broadinstitute/dsde/rawls/user/UserServiceSpec.scala:777]: Consider testing the addition of the azure billing project (when enabled).

**Review by cahrens — APPROVED** [review]
Overall this LGTM

File: core/src/main/scala/org/broadinstitute/dsde/rawls/billing/BillingProfileManagerDAO.scala (hunk 6 above)
Comment (aherbst-broad) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/billing/BillingProfileManagerDAO.scala]: Answer: yes, in my testing as long as the sam resources are properly setup we can add owners + users.

File: core/src/test/scala/org/broadinstitute/dsde/rawls/billing/BillingProfileManagerDAOSpec.scala [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+      SamRolesAndActions(
+        Set(SamBillingProjectRoles.owner),
+        Set(SamBillingProjectActions.createWorkspace)
+      ),
+      SamRolesAndActions(Set.empty, Set.empty),
+      SamRolesAndActions(Set.empty, Set.empty),
+      Set.empty,
+      Set.empty
+    )
+    when(samDAO.listUserResources(SamResourceTypeNames.billingProject, userInfo)).thenReturn(
```
Comment (marctalbott) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/billing/BillingProfileManagerDAOSpec.scala]: can we add a Google billing project to the return value to make sure the filter is working correctly?

**Review by marctalbott — APPROVED** [review]
Swagger needs to be updated as well to include the optional ManagedAppCoordinates in the billing project response

File: core/src/test/scala/org/broadinstitute/dsde/rawls/user/UserServiceSpec.scala (hunk 3 above)
Comment (aherbst-broad) [reply core/src/test/scala/org/broadinstitute/dsde/rawls/user/UserServiceSpec.scala:777]: Done in https://github.com/broadinstitute/rawls/pull/1690/commits/02e527dbdae779c9eab3db7b4c0724b6d7254828

**Comment by aherbst-broad** [comment]
jenkins retest

**Pull request merged**